third_party/move/move-compiler-v2/src/pipeline/variable_coalescing.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements a transformation that reuses locals of the same type when
+//! possible.
+//! prerequisite: livevar annotation is available by performing liveness analysis.
+//! side effect: this transformation removes all pre-existing annotations.
+//!
+//! This transformation is closely related to the register allocation problem in
+//! compilers. As such, an optimal solution to reusing locals is NP-complete, as we
+//! can show its equivalence to the graph coloring problem.
+//!
+//! Our solution here is inspired by the paper "Linear Scan Register Allocation"
+//! by Poletto and Sarkar, which proposes a fast and greedy register allocation
+//! algorithm. While potentially suboptimal, it is simple and fast, and is known to
+//! produce good results in practice. Our solution uses some key ideas from that
+//! paper, and performs a linear scan for deciding which locals to reuse.
+//!
+//! A key concept in this transformation is the "live interval" of a local, as opposed
+//! to the more precise "live range" (the set of code offsets where a local is live).
+//! The live interval of a local `t` is a consecutive range of code offsets `[i, j]`
+//! such that there is no code offset `j' > j` where `t` is live at `j'`, and there is
+//! no code offset `i' < i` where `t` is live at `i'`. A trivial live interval for any
+//! local is `[0, MAX_CODE_OFFSET]`, but we can often compute more precise live intervals.
+//!
+//! The transformation greedily reuses (same-typed) locals outside their live intervals.
+
+use crate::pipeline::livevar_analysis_processor::LiveVarAnnotation;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv, ty::Type};
+use move_stackless_bytecode::{
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// The live interval of a local (inclusive).
+/// Note that two live intervals i1: [b1, x] and i2: [x, e2] are *not* considered to overlap
+/// even though the code offset `x` is in both intervals.
+struct LiveInterval {
+    begin: CodeOffset,
+    end: CodeOffset,
+}
+
+impl LiveInterval {
+    /// Create a new live interval that only has the given offset.
+    fn new(offset: CodeOffset) -> Self {
+        Self {
+            begin: offset,
+            end: offset,
+        }
+    }
+
+    /// Include the given offset in the live interval, expanding the interval as necessary.
+    fn include(&mut self, offset: CodeOffset) {
+        self.begin = std::cmp::min(self.begin, offset);
+        self.end = std::cmp::max(self.end, offset);
+    }
+}
+
+/// Live interval event of a local, used for sorting.
+enum LiveIntervalEvent {
+    Begin(
+        /* which local? */ TempIndex,
```

Comment:
These comments are pretty unnecessary, except perhaps "live variable length", which should be a RHS comment:
```
    Begin(
        TempIndex,
        CodeOffset,
        usize, // live variable length, used for tie-breaking
    ),
    End(TempIndex, CodeOffset),
```

Reply:
One other way to comment this (and make it accessible to docgen) is:

```
    /// `Begin(local, start, len)` describes the event where `local` <blah>
    Begin(TempIndex, CodeOffset, usize)
```

Reply:
done